core/tests/extension.rs

```diff
@@ -0,0 +1,129 @@
+use lv2_core::extension::Extension;
+use lv2_core::lv2_extensions;
+use lv2_core::plugin::{lv2_descriptors, Plugin, PluginInfo, PortContainer};
+use lv2_core::uri::UriBound;
+use std::any::Any;
+use std::os::raw::c_void;
+
+// Test extensions
+
+trait FooExtension {
+    fn foo(&self) -> f32;
+}
+
+#[derive(Copy, Clone)]
+struct LV2FooInterface {
+    // This is the _sys crate usually
+    foo: unsafe extern "C" fn(handle: *mut c_void) -> f32,
+}
+
+unsafe extern "C" fn foo_ext_impl<P: FooExtension>(handle: *mut c_void) -> f32 {
+    (&*(handle as *const P)).foo()
+}
```

Comment:
This function should be already private in a module implementing an LV2 extension, so no `#[doc(hidden)]` is needed here, although I agree that the this isn't clear at all for this test.

I reorganized the contents of this test to create two example modules (which would be crates implementing fictional LV2 specifications): `lv2_foo` and `lv2_bar`. The only public things from these modules are the extension trait, and the interface struct in their `-sys` crate.
I hope this makes things clearer about what is accessible to the end-user, and what is not. :slightly_smiling_face:

Reply:
I still think that this method should be inside the trait, for example because we would need one generic parameter less. It would also be easier to identify a safe trait method with it's unsafe counter-part.

Reply:
The issue with putting it in the trait is that it makes it callable by the end user (which should not happen, as it is a callback for the host to call), and worse, it makes it override-able by the user, which is not something that is wanted by the implementer of the extension trait, as this function is the one that takes care of setting up all the necessary abstractions, and changing it is pretty much guaranteed to be incorrect.

This function (`foo_ext_impl`) is not an unsafe counterpart of the function to be implemented in the trait, it is the callback given to the host that will handle parameter conversion, unwind catching, error management, etc. It should not be accessible to the end-user by any means.

Reply:
This callback method is `unsafe`, if it is placed in the trait or not, and if we put it in the trait, it will also be hidden from the docs. This means, that the users have to a) discover that this callback exists, which they can only do if they read the source code, and b) write `unsafe` code to call or override this method, which tells them that **they have to know what they are doing**. If the users know what they are doing, they know that they should not call or override this method, because the (hidden) documentation in the source code tells them so. But if they do override the callback, which damage could be done? There will always be *a* callback function and we will therefore never have an invalid callback function pointer. The only thing I can think of is misuse of the raw pointers, but since the users are writing `unsafe` code, we have to trust them that they do proper work; That's one meaning of `unsafe` code: You always have to trust that `unsafe` code is correct.

On the other hand, as I said before, placing this method in the trait makes the extension code easier to read, because you know which callback belongs to which method and you have no type parameters. From my point of view, you want to hide too much from the users, which I disagree with: These kind of things should be exposed as `unsafe` code, which means that it is "hidden" to the normal user, but exposed for those who need that extra bit of control.

Reply:
I fail to see the use case for overriding the "entry point" host callback, as it basically overwrites the whole abstraction done by the implementer of the Extension (they would need to completely rewrite the whole parameter conversion, panic catching, thread-related invariants, etc.), basically monkey-patching a core function like this feels very hacky to me. But I see your point in general, extra control may be needed at times.

However, I think "how much should the extension expose to the user" is a question only the implementer of said extension should ask themselves, not us (as of making the extension system as a whole). The current system allows the callback to be either inside or outside of the trait, and if they keep it outside they can choose to expose any selection of (safe or unsafe) details to the user by adding that many methods on the trait for the end user, so i can see both points of view cohabiting in the ecosystem.

And if the end user disagrees with the implementer (and don't want to submit a bug report/PR for some reason), they can also write the extension in the way they want. After all, you can import as many implementations of a same extension (or port, or feature) as you want, as long as you only use one for each plugin. :slightly_smiling_face: 

As for this specific case, I don't think `Foo` or `Bar` don't need extra control as they are just tests. :wink: However, I will make a page explaining how to implement an extension, and what are the different ways you can expose more or less stuff to the user.

(Also, I really dislike exposing unsafe mehods with `#[doc(hidden)]`. If they are in the trait, they are publicly visible, so they should be properly documented. Otherwise it would only cause confusion to users as they could see it in the code but not access it in the docs.)

Reply:
I also see no use case for it, but I also don't see that it could do any harm that regular `unsafe` doesn't do. The mean reason I keep discussing is that the style of this example/test will heavily influence upcoming extensions, which we will have to implement too! Therefore, this example/test has to follow a style that is suitable for upcoming, bigger extensions and I think that keeping the callback functions separate from the trait will make the code messy.